container/docker/docker_test.go

```diff
@@ -0,0 +1,237 @@
+				"total_pgfault":97911,
+				"total_pgmajfault":165,
+				"total_pgpgin":66198,
+				"total_pgpgout":52876,
+				"total_rss":1048576,
+				"total_rss_huge":0,
+				"total_unevictable":0,
+				"total_writeback":0,
+				"unevictable":0,
+				"writeback":0},
+				"limit":2084458496
+			},
+			"name":"/test_container2",
+			"id":"container_id2",
+			"networks":{}
+		}`
+)
+
+// createContainers creates and starts the test containers, given a list of container names.
+func createContainers(ctx context.Context, cli *client.Client, containerNames []string) error {
+	for _, cname := range containerNames {
+		// create the docker container
+		cbody, err := cli.ContainerCreate(ctx, &cn.Config{Image: "ubuntu", Tty: true}, &cn.HostConfig{},
+			&nw.NetworkingConfig{}, cname)
+		if err != nil {
+			return err
+		}
+
+		// start the docker container
+		err = cli.ContainerStart(ctx, cbody.ID, types.ContainerStartOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// removeContainers removes all the containers given a list of containers as a part of clean up step.
+func removeContainers(ctx context.Context, cli *client.Client, containerList []*container.Container) error {
+	for _, cntr := range containerList {
+		err := cli.ContainerRemove(ctx, cntr.ID, types.ContainerRemoveOptions{Force: true})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func TestDockerFuncs(t *testing.T) {
+	ctx := context.Background()
+
+	cli, err := client.NewEnvClient()
+	if err != nil {
+		t.Skip("skipping test; new docker client cannot be created.")
+	}
+
+	_, err = cli.Ping(ctx)
+	if err != nil {
+		t.Skip("skipping test; docker deamon is not running")
+	}
+
+	// with a nonempty client
+	dockerOperator := NewInteractor(cli)
+
+	// with an empty client
+	dockerEmptyOperator := NewInteractor(&client.Client{})
+
+	// create and start the test containers
+	var cntrNames = []string{"container_1", "container_2"}
+
+	err = createContainers(ctx, cli, cntrNames)
+	if err != nil {
+		t.Skip("skipping tests; containers could not be created.")
+	}
+
+	Convey("Interactor implements container.Interactor", t, func() {
+		var _ container.Interactor = (*Interactor)(nil)
+	})
+
+	Convey("Decode the Container stats", t, func() {
+		Convey("for empty ReaderCloser stats", func() {
+			emptyRC := ioutil.NopCloser(bytes.NewReader([]byte("")))
+			emptyReaderCloserStats := types.ContainerStats{Body: emptyRC, OSType: "linux"}
+
+			stats, err := decodeDockerContainerStats(emptyReaderCloserStats)
+			So(stats, ShouldBeNil)
+			So(err, ShouldNotBeNil)
+		})
+
+		Convey("for non-empty ReaderCloser stats", func() {
+			nonEmptyRC := ioutil.NopCloser(bytes.NewReader([]byte(testReaderCloserStats)))
+			nonEmptyReaderCloserStats := types.ContainerStats{Body: nonEmptyRC, OSType: "linux"}
+
+			stats, err := decodeDockerContainerStats(nonEmptyReaderCloserStats)
+			So(stats, ShouldNotBeNil)
+			So(err, ShouldBeNil)
+		})
+	})
+
+	Convey("Given a Docker Operator", t, func() {
```

Comment:
dockerOperator and dockerEmptyOperator should be created in here, so each sub test starts with a known new fresh one that prior tests haven't interfered with.

Reply:
Done